File: Leetcode/C++/TwoPointer/maxArea.cpp
/*
Solution: 
- 双指针
- 定义指针i,j作为容器的左右边界， 容器的面积是(j - i) * min(height[i], height[j]);
- 下面介绍两种方法：

Solution 1：时间复杂度 O(n^2)
- j遍历数组，i在[0,j]内进行遍历；

Solution 2：时间复杂度 O(n)
- 要使得容器面积大，希望宽度和高度度都尽量大；
- 初始时，i,j分别放在数组的首尾，此时容器宽度最大；
- 因为h = min(height[i], height[j])，所以进入height[i] <= h这个while循环的条件就是，h=height[i],
  面积就是(j - i) * height[i]，那么就要尽量去找比较大的height[i],所以i++;
  同理，当h=height[j]时；
- 如果height[i] == height[j], 两个while循环都会进入，这个不会影响，因为最大面积边界要么是[i,j],要么在[i+1, j-1]之内；
  不可能是[i+1,j]或者[i,j+1],这两种情况的面积都小于[i,j];
*/
//  Solution 1
class Solution {
public:
    int maxArea(vector<int>& height) {
        int container = 0, i = 0, j = 1;
        while(j < height.size()){
            i = 0;
            while(i < j){
                int temp = (j - i) * min(height[i], height[j]);
                container = max(container, temp);
                i++;
            }
            j++;
        }
        return container;
    }
};


// Solution 2
class Solution {
public:
    int maxArea(vector<int>& height) {
        int container = 0, i = 0, j = height.size()-1;
        while(i < j){
            int h = min(height[i], height[j]);
            container = max(container, (j-i)*h);
            while(height[i] <= h and i < j){
                i++;
            }
            while(height[j] <= h and i < j){
                j--;
            }
        }
        return container;        
    }
};

File: Leetcode高频题/maxArea.py
'''
@lsy 2019.10.27

最大面积取决于两个边界中较小的一边。

Solution 1:
暴力，brute force， TLE(Time Limited Error)
i从0开始，j从i+1开始，找那个最大的面积即可。
时间复杂度O(n^2)

Solution 2:
双指针。
从两边往中间找，移动较小的那个边界。
时间复杂度O(n)
'''
# Solution 1
class Solution:
    def maxArea(self, height: List[int]) -> int:
        res = 0
        for i in range(len(height)):
            for j in range(i+1, len(height)):
                res = max(res, min(height[i], height[j])*(j-i))
        return res

# Solution 2
class Solution:
    def maxArea(self, height: List[int]) -> int:
        res = 0
        i, j = 0, len(height) - 1
        while i < j:
            res = max(res, min(height[i], height[j]) * (j-i))
            if height[i] < height[j]:
                i += 1
            else:
                j -= 1
        return res